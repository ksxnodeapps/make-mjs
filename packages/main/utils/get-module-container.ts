import { iterateAncestorDirectories, joinPath } from '@make-mjs/utils'

export function * getModuleContainer (path: string, basename = 'node_modules') {
  for (const dir of iterateAncestorDirectories(path)) {
    yield joinPath(dir, basename)
  }
}

export default getModuleContainer
