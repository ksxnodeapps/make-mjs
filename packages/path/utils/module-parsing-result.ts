export enum ModulePathKind {
  Internal = 'Internal',
  External = 'External'
}

export interface ModulePathParsingResultBase {
  readonly kind: ModulePathKind
  readonly name: string | null
  readonly path: string
}

export interface InternalModulePathParsingResult extends ModulePathParsingResultBase {
  readonly kind: ModulePathKind.Internal
  readonly name: null
}

export interface ExternalModulePathParsingResult extends ModulePathParsingResultBase {
  readonly kind: ModulePathKind.External
  readonly name: string
}

export type ModulePathParsingResult =
  InternalModulePathParsingResult |
  ExternalModulePathParsingResult
