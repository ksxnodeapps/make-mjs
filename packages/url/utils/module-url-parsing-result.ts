export enum ModuleUrlKind {
  Internal = 'Internal',
  External = 'External'
}

export interface ModulePathParsingResultBase {
  readonly kind: ModuleUrlKind
  readonly name: string | null
  readonly path: string
}

export interface InternalModuleUrlParsingResult extends ModulePathParsingResultBase {
  readonly kind: ModuleUrlKind.Internal
  readonly name: null
}

export interface ExternalModuleUrlParsingResult extends ModulePathParsingResultBase {
  readonly kind: ModuleUrlKind.External
  readonly name: string
}

export type ModuleUrlParsingResult =
  InternalModuleUrlParsingResult |
  ExternalModuleUrlParsingResult
