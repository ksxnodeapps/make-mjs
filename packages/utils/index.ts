export type MaybePromise<X> = X | Promise<X>

export const REJECTION_SILENCER = (): void => undefined

/**
 * Silence a unhandled promise rejection warning
 * @param promise Promise to silence
 */
export function silenceRejection<X> (promise: Promise<X>) {
  void promise.catch(REJECTION_SILENCER)
  return promise
}

/**
 * Remove trailing `'/'` if there is one
 * @param path Path to be rid of trailing separator
 */
export function removeTrailingSeparator (path: string) {
  return path.endsWith('/')
    ? path.slice(0, -1)
    : path
}

/**
 * Like `path.join` but does not eliminate `'.'`
 * @param paths Paths to join together
 */
export function joinPath (...paths: string[]) {
  return paths
    .map(removeTrailingSeparator)
    .join('/')
}
