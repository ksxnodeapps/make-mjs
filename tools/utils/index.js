require('@tools/preloaded-node/register')
const lib = require('./lib')
module.exports = lib.esm(lib)
