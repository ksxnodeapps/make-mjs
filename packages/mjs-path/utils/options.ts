import { MaybePromise } from '@make-mjs/utils'

export type FullMjsPathOptions = Required<MjsPathOptions>

export interface ModulePathResolver {
  (options: FullMjsPathOptions): MaybePromise<string>
}

export interface ModulePathTester {
  (options: FullMjsPathOptions): MaybePromise<boolean>
}

export interface MjsPackageTesterOptions extends MjsPathOptions {
  readonly packageName: string
  readonly moduleEntry: string
}

export interface MjsPackageTester {
  (options: MjsPackageTesterOptions): MaybePromise<boolean>
}

export interface MjsPathOptions {
  readonly modulePath: string
  readonly moduleContainer: string
  readonly fromFile?: ModulePathResolver
  readonly fromDir?: ModulePathResolver
  readonly isMjsPath?: ModulePathTester
  readonly isInternalModule?: ModulePathTester
  readonly isMjsPackage?: MjsPackageTester
}

export interface FromDirOptions extends MjsPathOptions {
  readonly modulePath: string
  readonly moduleContainer: string
  readonly isMjsPackage: MjsPackageTester
  readonly newExt?: string
}
