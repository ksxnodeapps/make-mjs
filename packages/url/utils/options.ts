import { MaybePromise } from '@make-mjs/utils'

export type FullMjsPathOptions = Required<MjsPathOptions>
export type FullModulePathResolverOptions = Required<ModulePathResolverOptions>

export interface ModuleUrlResolver {
  (options: ModulePathResolverOptions): MaybePromise<string>
}

export interface ModulePathResolverOptions extends MjsPathOptions {
  readonly modulePath: string
  readonly moduleContainer: readonly string[]
  readonly isMjsPackage: MjsPackageTester
  readonly forceMjs: boolean
  readonly newExt?: string
}

export interface ModuleUrlTester {
  (options: FullMjsPathOptions): MaybePromise<boolean>
}

export interface MjsPackageTesterOptions extends MjsPathOptions {
  readonly packageName: string
}

export interface MjsPackageTester {
  (options: MjsPackageTesterOptions): MaybePromise<boolean>
}

export interface MjsPathOptions {
  readonly modulePath: string
  readonly moduleContainer: readonly string[]
  readonly fromFile?: ModuleUrlResolver
  readonly fromDir?: ModuleUrlResolver
  readonly isMjsUrl?: ModuleUrlTester
  readonly isInternalModule?: ModuleUrlTester
  readonly isMjsPackage?: MjsPackageTester
}
