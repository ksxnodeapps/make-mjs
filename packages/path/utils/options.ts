import { MaybePromise } from '@make-mjs/utils'
import { ModulePathParsingResult } from './module-parsing-result'

export type FullMjsPathOptions = Required<MjsPathOptions>
export type FullModulePathResolverOptions = Required<ModulePathResolverOptions>

export interface ModulePathResolver {
  (options: ModulePathResolverOptions): MaybePromise<string>
}

export interface ModulePathResolverOptions extends MjsPathOptions {
  readonly modulePath: string
  readonly moduleContainer: string
  readonly isMjsPackage: MjsPackageTester
  readonly modulePathParsingResult: ModulePathParsingResult
  readonly forceMjs: boolean
  readonly newExt?: string
}

export interface ModulePathTester {
  (options: FullMjsPathOptions): MaybePromise<boolean>
}

export interface MjsPackageTesterOptions extends MjsPathOptions {
  readonly packageName: string
}

export interface MjsPackageTester {
  (options: MjsPackageTesterOptions): MaybePromise<boolean>
}

export interface MjsPathOptions {
  readonly modulePath: string
  readonly moduleContainer: string
  readonly fromFile?: ModulePathResolver
  readonly fromDir?: ModulePathResolver
  readonly isMjsPath?: ModulePathTester
  readonly isInternalModule?: ModulePathTester
  readonly isMjsPackage?: MjsPackageTester
}
