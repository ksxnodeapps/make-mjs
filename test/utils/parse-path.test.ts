import { parsePath } from '@make-mjs/utils'

describe('"abc/def/ghi/jkl.mno"', () => {
  const path = 'abc/def/ghi/jkl.mno'
  const get = () => parsePath(path)

  it('matches snapshot', () => {
    expect(get()).toMatchSnapshot()
  })

  it('has "dir" property', () => {
    expect(get()).toHaveProperty('dir', 'abc/def/ghi')
  })

  it('has "base" property', () => {
    expect(get()).toHaveProperty('base', 'jkl.mno')
  })

  it('has "name" property', () => {
    expect(get()).toHaveProperty('name', 'jkl')
  })

  it('has "ext" property', () => {
    expect(get()).toHaveProperty('ext', '.mno')
  })
})

describe('"abc/def/ghi/jkl.mno/"', () => {
  const path = 'abc/def/ghi/jkl.mno/'
  const get = () => parsePath(path)

  it('matches snapshot', () => {
    expect(get()).toMatchSnapshot()
  })

  it('has "dir" property', () => {
    expect(get()).toHaveProperty('dir', 'abc/def/ghi')
  })

  it('has "base" property', () => {
    expect(get()).toHaveProperty('base', 'jkl.mno')
  })

  it('has "name" property', () => {
    expect(get()).toHaveProperty('name', 'jkl')
  })

  it('has "ext" property', () => {
    expect(get()).toHaveProperty('ext', '.mno')
  })
})

describe('"abc/def/ghi/jkl"', () => {
  const path = 'abc/def/ghi/jkl'
  const get = () => parsePath(path)

  it('matches snapshot', () => {
    expect(get()).toMatchSnapshot()
  })

  it('has "dir" property', () => {
    expect(get()).toHaveProperty('dir', 'abc/def/ghi')
  })

  it('has "base" property', () => {
    expect(get()).toHaveProperty('base', 'jkl')
  })

  it('has "name" property', () => {
    expect(get()).toHaveProperty('name', 'jkl')
  })

  it('has "ext" property', () => {
    expect(get()).toHaveProperty('ext', '')
  })
})

describe('"abc/def/ghi/jkl."', () => {
  const path = 'abc/def/ghi/jkl.'
  const get = () => parsePath(path)

  it('matches snapshot', () => {
    expect(get()).toMatchSnapshot()
  })

  it('has "dir" property', () => {
    expect(get()).toHaveProperty('dir', 'abc/def/ghi')
  })

  it('has "base" property', () => {
    expect(get()).toHaveProperty('base', 'jkl.')
  })

  it('has "name" property', () => {
    expect(get()).toHaveProperty('name', 'jkl')
  })

  it('has "ext" property', () => {
    expect(get()).toHaveProperty('ext', '.')
  })
})

describe('"/abc.def"', () => {
  const path = '/abc.def'
  const get = () => parsePath(path)

  it('matches snapshot', () => {
    expect(get()).toMatchSnapshot()
  })

  it('has "dir" property', () => {
    expect(get()).toHaveProperty('dir', '')
  })

  it('has "base" property', () => {
    expect(get()).toHaveProperty('base', 'abc.def')
  })

  it('has "name" property', () => {
    expect(get()).toHaveProperty('name', 'abc')
  })

  it('has "ext" property', () => {
    expect(get()).toHaveProperty('ext', '.def')
  })
})

describe('"abc.def"', () => {
  const path = 'abc.def'
  const get = () => parsePath(path)

  it('matches snapshot', () => {
    expect(get()).toMatchSnapshot()
  })

  it('has "dir" property', () => {
    expect(get()).toHaveProperty('dir', '')
  })

  it('has "base" property', () => {
    expect(get()).toHaveProperty('base', 'abc.def')
  })

  it('has "name" property', () => {
    expect(get()).toHaveProperty('name', 'abc')
  })

  it('has "ext" property', () => {
    expect(get()).toHaveProperty('ext', '.def')
  })
})

describe('"/abc.def"', () => {
  const path = '/abc.def'
  const get = () => parsePath(path)

  it('matches snapshot', () => {
    expect(get()).toMatchSnapshot()
  })

  it('has "dir" property', () => {
    expect(get()).toHaveProperty('dir', '')
  })

  it('has "base" property', () => {
    expect(get()).toHaveProperty('base', 'abc.def')
  })

  it('has "name" property', () => {
    expect(get()).toHaveProperty('name', 'abc')
  })

  it('has "ext" property', () => {
    expect(get()).toHaveProperty('ext', '.def')
  })
})

describe('""', () => {
  const path = ''
  const get = () => parsePath(path)

  it('matches snapshot', () => {
    expect(get()).toMatchSnapshot()
  })

  it('has "dir" property', () => {
    expect(get()).toHaveProperty('dir', '')
  })

  it('has "base" property', () => {
    expect(get()).toHaveProperty('base', '')
  })

  it('has "name" property', () => {
    expect(get()).toHaveProperty('name', '')
  })

  it('has "ext" property', () => {
    expect(get()).toHaveProperty('ext', '')
  })
})
