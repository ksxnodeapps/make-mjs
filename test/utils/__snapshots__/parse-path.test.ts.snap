// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"" matches snapshot 1`] = `
Object {
  "base": "",
  "dir": "",
  "ext": "",
  "name": "",
}
`;

exports[`"/abc.def" matches snapshot 1`] = `
Object {
  "base": "abc.def",
  "dir": "",
  "ext": ".def",
  "name": "abc",
}
`;

exports[`"/abc.def" matches snapshot 2`] = `
Object {
  "base": "abc.def",
  "dir": "",
  "ext": ".def",
  "name": "abc",
}
`;

exports[`"abc.def" matches snapshot 1`] = `
Object {
  "base": "abc.def",
  "dir": "",
  "ext": ".def",
  "name": "abc",
}
`;

exports[`"abc/def/ghi/jkl" matches snapshot 1`] = `
Object {
  "base": "jkl",
  "dir": "abc/def/ghi",
  "ext": "",
  "name": "jkl",
}
`;

exports[`"abc/def/ghi/jkl." matches snapshot 1`] = `
Object {
  "base": "jkl.",
  "dir": "abc/def/ghi",
  "ext": ".",
  "name": "jkl",
}
`;

exports[`"abc/def/ghi/jkl.mno" matches snapshot 1`] = `
Object {
  "base": "jkl.mno",
  "dir": "abc/def/ghi",
  "ext": ".mno",
  "name": "jkl",
}
`;

exports[`"abc/def/ghi/jkl.mno/" matches snapshot 1`] = `
Object {
  "base": "jkl.mno",
  "dir": "abc/def/ghi",
  "ext": ".mno",
  "name": "jkl",
}
`;
